File: 309.best-time-to-buy-and-sell-stock-with-cooldown.py
from typing import List


# @leet start
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # buy at lowest price
        # sell at highest price
        # take 0 to i as the best result after selling everything
        # at time i, you can buy exactly once, and sell at time i-1.
        #       or not buy at all
        # at time j, you take the best result at time i,
        #       then consider i + 1 to j

        dp = dict()
        n = len(prices)

        def dfs(i: int, canBuy: bool) -> int:
            nonlocal dp

            if i >= n:
                return 0
            if (i, canBuy) in dp:
                return dp[(i, canBuy)]

            cooldown = dfs(i + 1, canBuy)
            if canBuy:
                buy = dfs(i + 1, False) - prices[i]
                dp[(i, canBuy)] = max(buy, cooldown)
            else:
                sell = dfs(i + 2, True) + prices[i]
                dp[(i, canBuy)] = max(sell, cooldown)

            return dp[(i, canBuy)]

        return dfs(0, True)


# @leet end

File: 309.best-time-to-buy-and-sell-stock-with-cooldown.cpp
#include <climits>
#include <vector>
using std::vector;

// @leet start
class Solution {
  public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int canBuy = 0;
        int canSell = -prices[0];
        int mustRest = INT_MIN;

        for (int i = 1; i < n; ++i) {
            // honestly, just make 3 temporary variables to avoid needing to
            // reorder things
            int tempRest = mustRest;
            mustRest = canSell + prices[i];
            canSell = std::max(canBuy - prices[i], canSell);
            canBuy = std::max(canBuy, tempRest);
        }
        return std::max(canBuy, mustRest);
    }

    // int maxProfit(vector<int>& prices) {
    //     int n = prices.size();
    //     vector<int> canBuy(n);
    //     vector<int> canSell(n);
    //     vector<int> mustRest(n);
    //     canBuy[0] = 0;
    //     canSell[0] = -prices[0];
    //     mustRest[0] = INT_MIN;
    //
    //     for (int i = 1; i < n; ++i) {
    //         canBuy[i] = std::max(canBuy[i - 1], mustRest[i - 1]);
    //         canSell[i] = std::max(canBuy[i - 1] - prices[i], canSell[i - 1]);
    //         mustRest[i] = canSell[i - 1] + prices[i];
    //     }
    //     return std::max(canBuy[n - 1], mustRest[n - 1]);
    // }
};
// @leet end
